### PR TITLE
feat(next/swc): support experimental swc plugins

### PR DESCRIPTION
**Pull request by kwonoj**
This PR enables experimental plugin support to `next/swc`. SWC-compatible plugins (i.e https://github.com/swc-project/plugins) can be loaded transparently into next-swc with new experimental configuration options.

Currently, next/swc enables multiple custom transform `@core/swc` doesn't support. It is based on using compile-time custom pass provided in swc's rust api. Users may not able to configure their own custom behavior or experiment dynamically unless core next.js include those features in the new release.

This feature in contrast, allows to configure runtime / dynamic way to specify custom transform behavior without rebuilding whole `next/swc` binary. New experiments, or user internal custom transformation which is not suitable for the core next.js can be enabled using this.

Configuring plugin is currently transparent to the core swc's config itself. It is likely possible `@swc/core` to have changes to the config interface as it's not clearly straightforward.

```js
// next.config.js
module.exports = {
  experimental: {
    swcPlugins: [
      ['plugin', {
        ..pluginOptions
      }]
    ]
  }
}
```

First element to the tuple points paths to the plugin - it can be either npm package name (if package.json's main field resolves to plugin binary) or absolute path the to `.wasm`. Second element is an option for configuring plugin.

Note wasm target (`next-swc/wasm`) does not support plugins yet: there is a tracking effort in upstream to support this. (https://github.com/swc-project/swc/issues/3934)

### Existing custom pass vs. plugin

As mentioned, `next/swc` already have numbers of custom transform passes overlaps to the plugin. Internally, both shares nearly identical visitor implementations - for example, https://github.com/vercel/next.js/pull/36692 uses custom pass but plugins for the `@swc/core` uses exact same visitor (https://github.com/kwonoj/swc-coverage-instrument/tree/main/packages/swc-coverage-instrument). The only difference is if it uses entrypoint for the plugin or not.

This means,
- `next/swc`, and core `next.js` may not need to change way to enable custom transform.
- Internal visitor implementation can be shared across, so _maybe_ next.js custom transform can be publihsed as plugin as well (which swc-project/plugins does for some already)
- Users for their application, or if some core experiment in next/swc needed without rebuilding whole, they can use plugins instead.

TL:DR plugins are more user-level, custom passes are for core next.js as-is.

Related to https://github.com/vercel/next.js/discussions/34384.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [x] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| buildDuration | 16.3s | 16.4s | ⚠️ +99ms |
| buildDurationCached | 6.4s | 6.3s | -104ms |
| nodeModulesSize | 611 MB | 611 MB | ⚠️ +562 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.981 | 3.963 | -0.02 |
| / avg req/sec | 628.02 | 630.8 | +2.78 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.394 | 1.367 | -0.03 |
| /error-in-render avg req/sec | 1793.84 | 1828.67 | +34.83 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.8 kB | 28.8 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.5 kB | 72.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.69 kB | 2.69 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.7 kB | 5.7 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.64 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 528 B | 528 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| buildDuration | 18.5s | 18.2s | -210ms |
| buildDurationCached | 6.3s | 6.3s | ⚠️ +17ms |
| nodeModulesSize | 611 MB | 611 MB | ⚠️ +562 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.018 | 4.024 | ⚠️ +0.01 |
| / avg req/sec | 622.18 | 621.3 | ⚠️ -0.88 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.403 | 1.393 | -0.01 |
| /error-in-render avg req/sec | 1781.33 | 1794.35 | +13.02 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.2 kB | 29.2 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 73.1 kB | 73.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.71 kB | 2.71 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.79 kB | 5.79 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.76 kB | 2.76 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| _buildManifest.js gzip | 456 B | 456 B | ✓ |
| Overall change | 456 B | 456 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | kwonoj/next.js feat-swc-experimental-plugin | Change |
| - | - | - | - |
| index.html gzip | 529 B | 529 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 2bbda2d1931f1f177f8b73e9937f942cdfbaa0d7

**Comment by kwonoj** [comment]
@Brooooooklyn may I able to ask some help to resolve CI failure? It passes native build CI job, but dependent tasks are failing to load binaries.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 09d6c084256022c414ff4773344b0f3496695dee

`yarn testheadless test/integration/telemetry/test/index.test.js` 
- Telemetry CLI > cli session: custom babel config (plugin)
- Telemetry CLI > emits telemetry for usage of swc

<details>
<summary>Expand output</summary>

  ● Telemetry CLI › cli session: custom babel config (plugin)

    ENOENT: no such file or directory, rename '/home/runner/work/next.js/next.js/test/integration/telemetry/.babelrc.plugin' -> '/home/runner/work/next.js/next.js/test/integration/telemetry/.babelrc'



  ● Telemetry CLI › emits telemetry for usage of swc

    ENOENT: no such file or directory, rename '/home/runner/work/next.js/next.js/test/integration/telemetry/next.config.swc' -> '/home/runner/work/next.js/next.js/test/integration/telemetry/next.config.js'



</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: docs/advanced-features/compiler.md [hunk 1]
```diff
@@ -316,6 +316,42 @@ This transform uses [handlebars](https://docs.rs/handlebars) to template the rep
 2. `member`: Has type `string`. The name of the member import.
 3. `lowerCase`, `upperCase`, `camelCase`: Helper functions to convert a string to lower, upper or camel cases.
 
+### SWC Trace profiling
```
Comment (padmaia) [inline docs/advanced-features/compiler.md:319]: Is this a separate feature? Thought I remembered a different PR for this one.

Comment (kwonoj) [reply docs/advanced-features/compiler.md:319]: Yes, it's a separate feature and forgot to update docs then included in this PR. If it's required, can split PR into another. (feature is already in, only docs are missing)

File: docs/advanced-features/compiler.md [hunk 2]
```diff
@@ -316,6 +316,42 @@ This transform uses [handlebars](https://docs.rs/handlebars) to template the rep
...
+    swcPlugins: [
+      ['plugin', {
+        ..pluginOptions
+      }]
+    ]
+  }
+}
+```
+
+`swcPlugins` accepts an array of tuple for configuring plugins. A tuple for the plugin contains path to the plugin, and an object for its configurations. Path to the plugin can be a npm module package name, or absolute path to the `.wasm` binary itself.
```
Comment (padmaia) [inline docs/advanced-features/compiler.md:353]:
```suggestion
`swcPlugins` accepts an array of tuples for configuring plugins. A tuple for the plugin contains the path to the plugin and an object for plugin configuration. The path to the plugin can be an npm module package name or an absolute path to the `.wasm` binary itself.
```

**Review by padmaia — APPROVED** [review]

**Comment by jantimon** [comment]
wow 🤩 thank you @padmaia & @kwonoj
